crates/next-error-code-swc-plugin/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+                let mut expr_iter = tpl.exprs.iter();
+
+                for (_i, quasi) in tpl.quasis.iter().enumerate() {
+                    result.push_str(&quasi.raw);
+                    if let Some(expr) = expr_iter.next() {
+                        result.push_str(&self.stringify_new_error_arg(expr));
+                    }
+                }
+                result
+            }
+
+            Expr::Bin(bin_expr) => {
+                // Assume binary expression is always add for two strings
+                format!(
+                    "{}{}",
+                    self.stringify_new_error_arg(&bin_expr.left),
+                    self.stringify_new_error_arg(&bin_expr.right)
+                )
+            }
+
+            _ => "%s".to_string(),
+        }
+    }
+
+    // Get error code, which is computed based on
+    // 1. Git commit hash, concatenated with
+    // 2. the hash of object { file_path, error_message, occurrence_count }
+    fn get_error_code(&mut self, first_arg: &ExprOrSpread) -> String {
+        let error_message: String = self.stringify_new_error_arg(&first_arg.expr);
+
+        // Maintain string_occurrences
+        *self
+            .string_occurrences
+            .entry(error_message.clone())
+            .or_insert(0) += 1;
+
+        // Generate the hash of object { file_path, error_message, occurrence_count }
+        let mut hasher = DefaultHasher::new();
+        serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string()
+        .hash(&mut hasher);
+        let hash = format!("{:08x}", hasher.finish());
+
+        // Concatenate the commit hash and the hash
+        let error_code = format!("E{}{}", self.commit_hash, hash);
+
+        // Save to JSON file
+        let error_metadata = serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string();
+        let mut retries = 3;
+
+        // Do not do FS operations in unit tests
+        if cfg!(test) {
+            return error_code;
+        }
+
+        loop {
+            if self.mode == "check" {
+                let error_code_path = format!("cwd/error_codes/{}.json", hash);
+                if !std::path::Path::new(&error_code_path).exists() {
+                    panic!(
+                        "ERROR: File {} does not exist.\n\nREQUIRED ACTION:\n1. Run `pnpm \
+                         build`\n2. Commit all file changes from \
+                         /packages/next/error-codes\n\nThis is required to maintain error code \
+                         consistency.",
+                        format!("/packages/next/error_codes/{}.json", hash)
+                    );
+                }
+                return error_code;
+            }
```

Comment:
AFAICT in `check` mode we either return or panic, nothing is retried, so i'd pull this out of the retry loop for readability